Adds ML supervised model Problem child package

alvarezmelissa87 opened the pull request:
## What does this PR do?

Adds the ML supervised model package for Problem child model.

Package includes:
- pipelines
- ml_module
- ml_model
- security rules

<img width="918" alt="ProblemChildCard2" src="https://user-images.githubusercontent.com/6446462/158696635-2e4157c4-492c-40f3-9fe7-eb2e2f7e516f.png">

<img width="1100" alt="ProblemChildReadme" src="https://user-images.githubusercontent.com/6446462/158696804-1c620b1e-d3d3-4bf1-9018-c79ffc6fd3b8.png">

<img width="1214" alt="problemChildOverview3" src="https://user-images.githubusercontent.com/6446462/158313194-3b9120f1-09a9-4e93-a70a-8462b77272c4.png">

<img width="1039" alt="ProblemChildNotice" src="https://user-images.githubusercontent.com/6446462/158696950-ad974a1c-e33b-4969-9a39-9f86cc71716c.png">


## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.
- [ ] If I'm introducing a new feature, I have modified the Kibana version constraint in my package's `manifest.yml` file to point to the latest Elastic stack release (e.g. `^7.13.0`).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
-

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2115/detail/PR-2115/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2115/detail/PR-2115/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2115/detail/PR-2115/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2115/detail/PR-2115/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_2115.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2021-11-02T23:53:43.255Z&rangeTo=2021-11-03T00:13:43.255Z&transactionName=Ingest-manager/integrations/PR-{number}&transactionType=unknown&latencyAggregationType=avg&traceId=0c1cc1f84bd9d7d8a05a4c2922bba715&transactionId=331c3a499ece1be3)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2021-11-03T00:03:43.255+0000


* Duration: 9 min 4 sec


* Commit: 9bd74bc3a907566da1533ac56844135e8df2fc97




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->











### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2115/detail/PR-2115/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Check integration: ml_problem_child`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-2115/runs/1/steps/521/log/?start=0">here</a></li>
<li>Description: <code>../../build/elastic-package check -v</code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-2115/runs/1/steps/557/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>


## :robot: GitHub comments

To re-run your PR in the CI, just comment with:


- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

peteharverson commented:
Is it possible to set a description field for the model, so that something like `The Problem child model is used to detect living off the land (LOtl) activity.` appears in the models list:

![image](https://user-images.githubusercontent.com/7405507/146572167-5c2e290d-c738-4b80-81a7-975113957fbd.png)

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,411 @@
+{
+  "attributes": {
+        "id": "problem_child",
+        "title": "Problem child anomaly detection",
```
I don't think we need `anomaly detection` in here - can just be `Problem child`.

alvarezmelissa87 replied:
Updated in 020da604273535260ba2dc54088c1bc9d63ad80f

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,411 @@
+  "attributes": {
+        "id": "problem_child",
+        "title": "Problem child anomaly detection",
+        "description": "Identify malicious processes.",
```
I'd suggest adding `detect potential living off the land activity` in here, something like this:

![image](https://user-images.githubusercontent.com/7405507/146575512-2b7b9c99-61da-484d-83fe-051ead08e81b.png)

alvarezmelissa87 replied:
Updated in 020da604273535260ba2dc54088c1bc9d63ad80f

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,411 @@
+      {
+        "id": "problem_child_rare_process_by_host",
+          "config": {
+            "description": "A machine learning job to detect potential living off the land activity. Looks for a process that has been classified as malicious on a host that does not commonly manifest malicious process activity. This is an experimental job and is therefore unsupported.",
```
I don't think we need `A machine learning job to detect potential living off the land activity.` at the start of all these descriptions. `A machine learning job` can definitely be removed, and probably the whole of the first sentence is unnecessary.

alvarezmelissa87 replied:
Updated in 020da604273535260ba2dc54088c1bc9d63ad80f

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,411 @@
+        "id": "datafeed-problem_child_rare_process_by_host",
+        "job_id": "problem_child_rare_process_by_host",
+        "config": {
+          "indices": [
```
The indices should not be hardcoded in the datafeed configs, or else it won't use the data view selected by the user in the ML job wizard. It should just be

```
  "indices": [
     "INDEX_PATTERN_NAME"
   ],
```

alvarezmelissa87 replied:
Updated in 020da604273535260ba2dc54088c1bc9d63ad80f

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,411 @@
+    "jobs": [
+      {
+        "id": "problem_child_rare_process_by_host",
+          "config": {
```
We probably want to add a `groups` property in here to make it easier to identify / search for these jobs in the UI - would need to find out what groups we should be adding for these jobs. `security`, `process` `problemchild` ?

![image](https://user-images.githubusercontent.com/7405507/146580711-75f733e9-af64-4f82-a5a1-196ab0dbe3d8.png)

peteharverson commented on packages/ml_problem_child/changelog.yml:
```diff
@@ -0,0 +1,6 @@
+# newer versions go on top
+- version: "0.0.1"
+  changes:
+    - description: Initial draft of the package
```
I think the word `draft` can be lost here - other packages I've looked at go with something like `Initial release`

dishadasgupta replied:
Agreed, I think draft implies this isn't ready for users to use

alvarezmelissa87 replied:
Updated in 7d33b9d65996d461f7d20ece7d0f96a636698584

peteharverson commented on packages/ml_problem_child/changelog.yml:
```diff
@@ -0,0 +1,6 @@
+  changes:
+    - description: Initial draft of the package
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/0 # FIXME Replace with the real PR link
```
You can put in the actual PR number now - `2115`

alvarezmelissa87 replied:
Updated in 7d33b9d65996d461f7d20ece7d0f96a636698584

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,459 @@
+        "id": "problem_child",
+        "title": "Problem child",
+        "description": "Detects potential living off the land activity by identifying malicious processes.",
+        "type": "ML model",
```
The `type` field in the module is generally used to describe the type of data the jobs run against. So maybe something like `Process data`  would be a better description here? Or even `Problem child`? Note this field isn't currently used in the UI - it just acts as an extra label in the module definition.

alvarezmelissa87 replied:
Updated in 7d33b9d65996d461f7d20ece7d0f96a636698584

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,459 @@
+                        "winlogbeat",
+                        "process"
+                    ],
+                    "description": "Looks for a process that has been classified as malicious on a host that does not commonly manifest malicious process activity. This is an experimental job and is therefore unsupported.",
```
Just want to confirm that we should be describing these jobs as experimental.

peteharverson replied:
Confirmed that yes, we should mark these jobs as experimental for now. The package too has the experimental tag on it.

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,459 @@
+                    "data_description": {
+                        "time_field": "@timestamp",
+                        "time_format": "epoch_ms"
+                    }
```
We should add `created_by` fields for all these jobs
```
  "custom_settings": {
    "created_by": "ml-module-problem-child"
  }
```

alvarezmelissa87 replied:
Added in 7d33b9d65996d461f7d20ece7d0f96a636698584

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,459 @@
+                        "detectors": [
+                            {
+                                "by_field_name": "host.hostname",
+                                "detector_description": "high sum by host",
```
Is there a reason why the two detectors in this job have been given the same description? Could it be given a more human readable description, such as `high number of malicious processes by host`?

ajosh0504 replied:
I second this.

alvarezmelissa87 replied:
Updated in 7d33b9d65996d461f7d20ece7d0f96a636698584

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,459 @@
+                            },
+                            {
+                                "by_field_name": "host.hostname",
+                                "detector_description": "high sum by host",
```
As above, could the description here be edited to indicate the field it is running on `blocklist_label`? I don't have enough context on what information is held in `blocklist_label` to advise here.

ajosh0504 replied:
@peteharverson So the way we classify processes as malicious is using either a supervised model or a blocklist for things the model may have missed. I think adding this in the description will introduce complexity.

peteharverson replied:
OK. Thanks for explanation @ajosh0504

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,459 @@
+                            },
+                            {
+                                "by_field_name": "user.name",
+                                "detector_description": "high sum by host",
```
Could we include the field name in the description of this detector - what is it a high sum of?

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,459 @@
+                        "detectors": [
+                            {
+                                "by_field_name": "user.name",
+                                "detector_description": "high sum by user",
```
Could we include the field name in the description of this detector - what is it a high sum of?

ajosh0504 replied:
As above, I think introducing field names in the descriptions might introduce complexity, since users may not know what the fields exactly mean.

dishadasgupta replied:
I agree with Apoorva, field names aren't always the most user-centric and don't add that much value (imo) for the complexity they could add

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,459 @@
+                        "detectors": [
+                            {
+                                "by_field_name": "process.parent.name",
+                                "detector_description": "high sum by parent process",
```
Could we include the field name in the description of this detector - what is it a high sum of?

peteharverson commented on packages/ml_problem_child/kibana/ml_module/problem-child-ml.json:
```diff
@@ -0,0 +1,459 @@
+                            },
+                            {
+                                "by_field_name": "process.parent.name",
+                                "detector_description": "high sum by host",
```
Could we include the field name in the description of this detector - what is it a high sum of?

peteharverson commented on packages/ml_problem_child/manifest.yml:
```diff
@@ -0,0 +1,24 @@
+title: "ML Problem Child"
+version: 0.0.1
+license: platinum
+description: "The Problem Child model is used to detect LOtL activity."
```
This package is more than just a 'model'. How about `The Problem Child framework is used to detect living of the land (LOtL) activity.`

alvarezmelissa87 replied:
Updated in 7d33b9d65996d461f7d20ece7d0f96a636698584

peteharverson commented on packages/ml_problem_child/manifest.yml:
```diff
@@ -0,0 +1,24 @@
+format_version: 1.0.0
+name: ml_problem_child
```
I'm not sure we need `ml` in the name, title and even the folder name (`ml_problem_child`) of the package, as the emphasis should be on the use case 'problem child'.

alvarezmelissa87 replied:
Updated in 7d33b9d65996d461f7d20ece7d0f96a636698584

ajosh0504 reviewed (commented):
Thanks for getting this done! Super excited to see these models in packages. Just one question. So once a user installs a package, what's the process for them to update/change artifacts produced by the package?

ajosh0504 commented:
As with the DGA model, should the model file name here be something more understandable? It's currently `problemchild_20210526_1.0.json`. This will also need to be changed in the inference pipeline.

dishadasgupta commented on packages/ml_problem_child/elasticsearch/ingest_pipeline/ml_problem_child_ingest_pipeline.yml:
```diff
@@ -0,0 +1,10 @@
+---
+description: "A pipeline of pipelines for ProblemChild detection"
```
Is there a way to make this description a bit clearer? I feel like "pipeline of pipelines" at first read could be a bit confusing

peteharverson replied:
How about just `Pipelines for ProblemChild detection`? Is this label actually shown anywhere in the UI?

dishadasgupta commented on packages/ml_problem_child/docs/README.md:
```diff
@@ -0,0 +1,6 @@
+
+The Problem child model package contains the [Problem child model and associated assets](https://www.elastic.co/blog/problemchild-generate-alerts-to-detect-living-off-the-land-attacks), which are used to detect living off the land (LOtl) activity.
+
+To download the assets, click **Settings** > **Install ML Problem child assets**.
```
I'm assuming this is within a cluster? Could we get an example of what that is supposed to look like? Are there any steps needed before this? i.e. ProblemChild requires Platinum level capabilities, is that denoted anywhere (unless I missed it)? Maybe even a quick note in the README would be helpful

alvarezmelissa87 replied:
The package manifest file has the license type that is required - that is enforced in Fleet. We could add a note in the README as well 👍

alvarezmelissa87 commented on packages/ml_problem_child/docs/README.md:
```diff
@@ -0,0 +1,6 @@
+# ML Problem Child
```
Updated to use `ProblemChild` everywhere in 7d33b9d65996d461f7d20ece7d0f96a636698584

alvarezmelissa87 commented on packages/ml_problem_child/docs/README.md:
```diff
@@ -0,0 +1,6 @@
+# ML Problem Child
+
+The Problem child model package contains the [Problem child model and associated assets](https://www.elastic.co/blog/problemchild-generate-alerts-to-detect-living-off-the-land-attacks), which are used to detect living off the land (LOtl) activity.
```
Updated in 7d33b9d65996d461f7d20ece7d0f96a636698584

alvarezmelissa87 commented:
cc @lcawl for package/model description text in the readme and such.

brokensound77 commented on packages/problem_child/kibana/security_rule/34184d4e-ef61-477b-8d76-5c93448c29bf.json:
```diff
@@ -0,0 +1,31 @@
+        "author": [
+            "Elastic"
+        ],
+        "description": "A supervised machine learning model (ProblemChild) or its blocklist has identified\na suspicious Windows process event to be malicious activity.\n",
```
is the line break in the middle intentional?

brokensound77 replied:
Shouldn't need a trailing one either

alvarezmelissa87 replied:
These lines breaks were part of the release and haven't been changed. I'd need confirmation that it's okay to remove them.

spong replied:
FWIW there doesn't appear to be any issue with rendering (these newlines are ignored):

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/153307527-04f086c7-2b80-4ffc-9a39-6a457d027e0e.png" />
</p>

brokensound77 commented on packages/problem_child/kibana/security_rule/9a2e372a-cbeb-4ad6-a288-017ef086324c.json:
```diff
@@ -0,0 +1,31 @@
+        "license": "Elastic License",
+        "max_signals": 10000,
+        "name": "Machine Learning Detected a Suspicious Windows Event with a High Malicious Probability Score",
+        "query": "(problemchild.prediction:1 and problemchild.prediction_probability \u003e 0.98) or blocklist_label:1",
```
was the `\u003e` a unicode conversion error?

alvarezmelissa87 replied:
Yep - fixed! thanks!

alvarezmelissa87 replied:
Thanks for catching this - this should now be a `>` instead of the encoded version 👍 Fixed in c3692a17d515ebdd31e95e03652f735d04db5748

alvarezmelissa87 replied:
@spong - looks like running the `elastic-package format` command is what's causing the change from `>` to the encoded version. Is this expected?
cc @mtojek

brokensound77 replied:
Oh wow, looks like you are right https://github.com/elastic/integrations/blob/1d18f55bd4e33f6d4b2c93bd2ad45e82bb1d4fd0/packages/security_detection_engine/kibana/security_rule/a9b05c3b-b304-4bf9-970d-acdfaef2944c.json#L16

I guess it is ok to leave it converted then 🤷

mtojek replied:
What we can see in the CI output is:

```
[2022-03-18T15:59:14.998Z] Error: checking package failed: formatting the integration failed (path: /var/lib/jenkins/workspace/est-manager_integrations_PR-2115/src/github.com/elastic/integrations/packages/problem_child, failFast: true): walking through the integration files failed: formatting file failed (path: /var/lib/jenkins/workspace/est-manager_integrations_PR-2115/src/github.com/elastic/integrations/packages/problem_child/kibana/security_rule/9a2e372a-cbeb-4ad6-a288-017ef086324c.json): file is not formatted (path: /var/lib/jenkins/workspace/est-manager_integrations_PR-2115/src/github.com/elastic/integrations/packages/problem_child/kibana/security_rule/9a2e372a-cbeb-4ad6-a288-017ef086324c.json)
```

so it confirms that you have to post the formatted code. Sometimes the formatted code might be harder to read, but we don't have a choice here if we want to depend on standard JSON libraries. Please run `elastic-package format` and post the formatted content.

mtojek replied:
> cc @mtojek - also - running elastic-package check locally is failing with a reference to the ml_model file name - though the update was made to package-spec to match that file name pattern and I've tested with the package-spec validator test so curious if this could be something going on locally? Might need to chat with you quickly on that one.

It depends on which elastic-package [release](https://github.com/elastic/elastic-package/releases) you have locally and which one is in the `go.mod` in the Integrations.

BTW we're struggling a bit with updating the dependency - [here](https://github.com/elastic/integrations/pull/2845).

brokensound77 reviewed (commented):
So cool to see this finally going into the product (after being in an experimental state for so long). I left a few suggestions.

@spong FYSA, if I am not mistaken, this and #2352 and the first integrations outside of the prebuilt rules to add rules destined for the detection engine. Not sure if there were any assumptions or considerations to be made.

brokensound77 commented on packages/problem_child/docs/README.md:
```diff
@@ -0,0 +1,6 @@
+The ProblemChild package contains the [ProblemChild model and associated assets](https://www.elastic.co/blog/problemchild-generate-alerts-to-detect-living-off-the-land-attacks), which are used to detect living off the land (LotL) activity.
+
+To download the assets, click **Settings** > **Install ML ProblemChild assets**.
+
```
is it worth elaborating on this to explain the relationship between the model, jobs, and rules. I believe the job and rules will be installed but in a disabled state. Also if the rule is enabled before the job, it will throw an error. 

The schema for rules accepts defining `enabled = true`, though I am not totally sure on the ml jobs

alvarezmelissa87 replied:
How about adding something to the `Configuration` section of the readme - after the `to download the assets` bit 
Something like `Ingest data with the installed ingest pipeline to enrich your indices with inference data and run the provided anomaly detection jobs.` 
cc @peteharverson

lcawl replied:
The goal is to ultimately have a page in the docs that walks through the installation and use of this integration's assets, but since that's not available to link to yet, I agree with your suggestion for a couple of lines in the Configuration section of the readme as a stop-gap.

alvarezmelissa87 replied:
Added in 11eb4a152ddadf13c75d6e469e97f01d6db9dd06
cc @lcawl, @peteharverson

alvarezmelissa87 replied:
@lcawl, @peteharverson, @Winterflower 
The latest screenshots and README file have been updated with more detailed instructions on how to use the package - as discussed - in c3692a17d515ebdd31e95e03652f735d04db5748
Would appreciate confirmation that it's what we want when you get a chance! 🙏

alvarezmelissa87 commented on packages/problem_child/manifest.yml:
```diff
@@ -0,0 +1,24 @@
+format_version: 1.0.0
+name: problem_child
+title: "ML ProblemChild"
```
Removed in 7d0d4520824c3e32aa600a4ecfa43d919e2217c2

peteharverson reviewed (approved):
Latest changes LGTM

alvarezmelissa87 commented:
This has been updated with all suggestions and is ready for another look when you get a chance 🙏 
cc @dishadasgupta, @ajosh0504, @brokensound77

spong commented on packages/problem_child/kibana/security_rule/34184d4e-ef61-477b-8d76-5c93448c29bf.json:
```diff
@@ -0,0 +1,31 @@
+        ],
+        "timestamp_override": "event.ingested",
+        "type": "query",
+        "version": 2
```
Since these are brand new rules is there a reason its initial version is `2`? Is there any versioning handled elsewhere by chance (perhaps auto-generated like they do in the `detection-rules` repo)?

alvarezmelissa87 replied:
cc @Winterflower, @dishadasgupta Not sure who I'd need to check with for the answer to this

brokensound77 replied:
My guess is that they were versioned based on their existence in pre-releases within the detection-rules repo (which was considered experimental and not fully supported). Best course of action would probably be to regenerate a new random UUID for `rule_id` and then reset to version 1

spong replied:
Not a big deal either way, these rules can ship with their initial version as `2` and it's no issue. There aren't many guidelines around versioning other than it must be incremented (if there's a previous version) for the update logic to catch it.

grabowskit commented:
Could we change the title of the card be "Living off the Land Attack Detection" or if there are character limits "LotL Attack Detection"?  Then include in the card description... "The ProblemChild package is used to detect Living off the Land attacks. Platinum license required" (having LotL spelled out in the description).  Also, let me know if we run into a character limit. I would like to keep ProblemChild in the card description because the ML model is called ProblemChild and the inference pipeline has ProblemChild description in it.

This would also require a change to the title after clicking on the card, but the Overview description looks like it won't need any other changes.

cc: @hmnichols

alvarezmelissa87 commented:
> Could we change the title of the card be "Living off the Land Attack Detection" or if there are character limits "LotL Attack Detection"? Then include in the card description... "The ProblemChild package is used to detect Living off the Land attacks. Platinum license required" (having LotL spelled out in the description). Also, let me know if we run into a character limit. I would like to keep ProblemChild in the card description because the ML model is called ProblemChild and the inference pipeline has ProblemChild description in it.
> 
> This would also require a change to the title after clicking on the card, but the Overview description looks like it won't need any other changes.
> 
> cc: @hmnichols

Updated the title card and the overview f05aa70beca85fb47b1a90c245097e288175ed18
Also updated the screenshots in the description

lcawl reviewed (approved):
One suggestion, otherwise text LGTM

ajosh0504 reviewed (approved):
LGTM if all the outstanding feedback has been resolved. 👍

bfilar reviewed (approved):
LGTM -- If anyone needs help with documentation please feel free to reach out to me.

dishadasgupta commented on packages/problem_child/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,39 @@
+# ProblemChild Detection Model
```
IMO, I think it'd be best to leave it as ProblemChild Detection Model for consistency for the user since it's being referenced as ProblemChild + the README goes on to describe the purpose

lcawl replied:
Won't that cause confusion? The title in the UI will not match the title in the docs (https://docs.elastic.co/en/integrations)

dishadasgupta replied:
@lcawl This link doesn't point to anything specific regarding ProblemChild that I can see, what will it be listed as in the UI? If it's not ProblemChild, perhaps that should change, since we've been calling it ProblemChild everywhere. Let me know if I'm misunderstanding something, thanks

lcawl replied:
Right, that documentation can only be created after this integration is pushed to the package repository and it's generated from whatever we put in the readme here. IMO the readme title should therefore match the integration's UI title.

alvarezmelissa87 replied:
Updated this to `LotL Attack Detection` in c3692a17d515ebdd31e95e03652f735d04db5748

dishadasgupta reviewed (approved):
LGTM once all outstanding comments are handled + checks pass. Likewise, I am around for any help with docs

lcawl commented on packages/problem_child/manifest.yml:
```diff
@@ -0,0 +1,24 @@
+description: "The ProblemChild framework is used to detect living off the land activity. Requires a platinum license."
+type: integration
+categories:
+  - custom
```
While testing just now I sought this package in the "security" category of integrations and didn't think to look in the "custom" category. If this is just a newbie mistake, feel free to ignore.

peteharverson replied:
I think we should make this `[custom, security]`. Or even just switch it to `security`?

alvarezmelissa87 replied:
Updated in c3692a17d515ebdd31e95e03652f735d04db5748

mtojek reviewed (changes requested):
Could you please update this PR with the latest main branch and add an appropriate entry to the CODEOWNERS file?

lcawl commented on packages/problem_child/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,59 @@
+
+To download the assets, click **Settings** > **Install LotL Attack Detection assets**.
+
+Follow these instructions to ingest data with the ingest pipeline and enrich your indices with inference data. Then use these detection rules and anomaly detection jobs to detect LotL attacks. For more detailed information refer to the [ProblemChild blog] (https://www.elastic.co/blog/problemchild-generate-alerts-to-detect-living-off-the-land-attacks)
```
I think this space causes a problematic link:
```suggestion
Follow these instructions to ingest data with the ingest pipeline and enrich your indices with inference data. Then use these detection rules and anomaly detection jobs to detect LotL attacks. For more detailed information refer to the [ProblemChild blog](https://www.elastic.co/blog/problemchild-generate-alerts-to-detect-living-off-the-land-attacks)
```

lcawl commented on packages/problem_child/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,59 @@
+
+Once you’ve installed the package you can ingest your data using the ingest pipeline. This will enrich your incoming data with its predictions from the machine learning model.
+
+This pipeline is designed to work with [Winglogbeat data] (https://www.elastic.co/downloads/beats/winlogbeat).
```
Ditto re unnecessary space:
```suggestion
This pipeline is designed to work with [Winglogbeat data](https://www.elastic.co/downloads/beats/winlogbeat).
```

lcawl commented on packages/problem_child/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,59 @@
+
+Create a data view for the indices that are enriched by the pipeline.
+
+In Machine Learning > Anomaly Detection, when you create a job, you should see an option to `Use preconfigured jobs` with a card for LotL Attacks. Select that option. Once you select the card you will see several pre-configured anomaly detection jobs that you can enable depending on what makes the most sense for your environment. Note these jobs are only useful for indices that have been enriched ...
```
Minor edits:
```suggestion
In **Machine Learning > Anomaly Detection**, when you create a job, you should see an option to `Use preconfigured jobs` with a card for LotL Attacks. When you select the card, you will see several pre-configured anomaly detection jobs that you can enable depending on what makes the most sense for your environment. Note these jobs are only useful for indices that have been enriched by the ingest pipeline.
```

lcawl commented on packages/problem_child/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,59 @@
+
+In Machine Learning > Anomaly Detection, when you create a job, you should see an option to `Use preconfigured jobs` with a card for LotL Attacks. Select that option. Once you select the card you will see several pre-configured anomaly detection jobs that you can enable depending on what makes the most sense for your environment. Note these jobs are only useful for indices that have been enriched ...
+
+### (Optional) Enable Security Rules
```
```suggestion
### (Optional) Enable security rules
```

lcawl commented on packages/problem_child/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,59 @@
+
+### (Optional) Enable Security Rules
+
+In order to maximize the benefit of the LotL Detection framework, you might consider activating detection rules that are triggered when certain conditions for the supervised model or anomaly detection jobs are satisfied. See the [documentation] (https://www.elastic.co/guide/en/security/current/detection-engine-overview.html for more information on importing and enabling the rules.
```
Ditto re unnecessary space and there's also a missing bracket:
```suggestion
In order to maximize the benefit of the LotL Detection framework, you might consider activating detection rules that are triggered when certain conditions for the supervised model or anomaly detection jobs are satisfied. See the [documentation](https://www.elastic.co/guide/en/security/current/detection-engine-overview.html) for more information on importing and enabling the rules.
```

alvarezmelissa87 commented:
This has been updated with the license/subscription language, the CODEOWNERS file has been updated, and the NOTICE.txt has been added. This is ready for a final look when you get a chance. cc @amandarichey, @mtojek, @spong

spong reviewed (approved):
LGTM from the Security Solution side. Thank you for all your efforts here and helping me get my test environment up and running @alvarezmelissa87! And congrats on the milestone here with the first ML integration! 🎉 🥳 🙌  

Only thing to note from the Security side is now that we have multiple integrations that are shipping assets of `"type": "security-rule"`, there could be the possibility of accidentally shipping rules that don't have a unique `rule_id` across all the other packages. This would manifest as an issue on the application side as the Security Solution app will only allow you to install rules with unique `rule_id`'s. I tested this locally by running this PR w/ package-registry and modifying one of the `rule_id`'s to be the same as one of the Detection Rule Integrations', and while both the integration asset SO's will be installed without issue, once you're in the Security UI it'll only install one of them (with no error or anything that there's a dupe, which is on our end, but is technically unsupported).

So that said, I think we need test case within the `integrations` repo here that checks any added assets of `"type": "security-rule"` and ensure's that there are no duplicates across all the other integrations. I'm not too familiar with integrations dev to know if this is possible, but would love to hear your thoughts if there may be another way of accomplishing this check. Since the rule asset `rule_id's` added in this PR are unique there's no need to hold up this PR, so we could just write an issue and handle this later (something my team @elastic/security-detections-response-rules can probably help with).

mtojek commented on packages/problem_child/manifest.yml:
```diff
@@ -0,0 +1,24 @@
+    size: 32x32
+    type: image/svg+xml
+owner:
+  github: elastic/integrations
```
Please change it to `elastic/ml-ui` as in the CODEOWNERS file.

alvarezmelissa87 replied:
Updated in f89096226aa2afbf8622cd7480b5712cd7cea54e

mtojek commented on .github/CODEOWNERS:
```diff
@@ -97,6 +97,7 @@
 /packages/panw @elastic/security-external-integrations
 /packages/pfsense @elastic/security-external-integrations
 /packages/postgresql @elastic/integrations
+/packages/problem_child @elastic/ml-ui
```
Please merge the latest main branch so that the CODEOWNERS file is fixed.

mtojek reviewed (changes requested):
I've added the `elastic/ml-ui` team "write" permissions to this repository.

jsoriano commented:
> I think we need test case within the `integrations` repo here that checks any added assets of `"type": "security-rule"` and ensure's that there are no duplicates across all the other integrations. I'm not too familiar with integrations dev to know if this is possible, but would love to hear your thoughts if there may be another way of accomplishing this check.

Take into account that not all packages are hosted in the `integrations` repo, there are packages whose source is in other repositories, and this is something we want to support. So we cannot trust only on checks on this repository.

There are other cases of assets that must have unique ids. For example for dashboards it is required that they are prefixed with the name of the package. Would this be an option here?

Other option could be that fleet guarantees that ids are unique on install time, I remember discussing this option for some asset in the past but I am not sure if this was implemented for some asset. The problem with this approach is that it can complicate maintaining references to these assets.

spong commented:
> Take into account that not all packages are hosted in the integrations repo, there are packages whose source is in other repositories, and this is something we want to support. So we cannot trust only on checks on this repository.

Thank you for clarifying @jsoriano! I wasn't aware, but kinda figured this might be the case (or at least that there'd be some support for arbitrary integrations hosted elsewhere).

> There are other cases of assets that must have unique ids. For example for dashboards it is required that they are prefixed with the name of the package. Would this be an option here?

Potentially, this `rule_id` essentially behaves as an external static id so that as these assets traverse different systems or continue to be reinstantiated we can still maintain a way of identifying them, so tagging them with their source integration shouldn't be disruptive here. That said we do have to take into consideration that we're already shipping `security-rule` assets in the `detection rules` integration, and those also separately ship pre-bundled with Kibana as well -- so would need to coordinate changes there too.

> Other option could be that fleet guarantees that ids are unique on install time, I remember discussing this option in the past but I am not sure if this was implemented for some asset. The problem with this approach is that it can complicate maintaining references to these assets.

I don't think this would be an option since as mentioned above we'll want this ID to remain static between installs/import/export/etc.

Let me discus the first option with our team and see if that'll work, or if folks have any other ideas here. Thanks again for the details here! 🙂

spong commented:
@jsoriano @alvarezmelissa87 

I discussed with the team this morning and the recommended output was to create a discussion issue with all our options and how it affects the UX and then discuss with Product this Thursday and let them decide -- so I'll be creating that issue later today and will share here in case there's any additional input.

I don't want to block this PR on these changes, but if they decide to prefix the rule_id's with the package name we'll have to ensure that these ID's are updated before this package is published. In talking with @alvarezmelissa87 I think we're fine to merge this when ready, but if we can wait till we hear back from Product on Thursday before publishing that would be great. 👍

mtojek commented on packages/problemchild/kibana/ml_module/problemchild-ml.json:
```diff
@@ -0,0 +1,477 @@
+            }
+        ]
+    },
+    "id": "problem_child",
```
@alvarezmelissa87 What about this "id"? CI mentioned that field.

alvarezmelissa87 replied:
Ah, thank you! I thought I changed it - updating 👍

alvarezmelissa87 replied:
Ah, thanks - missed this the first time! Updated in 38fd11b82218701d16107b60c6e418b8f7c9ff2f

spong commented:
Just commenting to update that we've come to consensus on how we'll handle the aforementioned issue around `rule_id` uniqueness between packages, and that there is no reason to hold up the merging or publishing of this package from the Security Solution side. Please see this comment https://github.com/elastic/kibana/issues/128202#issuecomment-1078457192 for all the details, and thank you all for all your input here! 🙂
